File: app/src/main/java/com/navtej/doordashlite/viewmodel/RestaurantViewModelFactory.java
package com.navtej.doordashlite.viewmodel;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;
import androidx.collection.ArrayMap;

import com.navtej.doordashlite.di.ViewModelSubComponent;

import java.util.Map;
import java.util.concurrent.Callable;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class RestaurantViewModelFactory implements ViewModelProvider.Factory {
    private final ArrayMap<Class, Callable<? extends ViewModel>> creators;

    @Inject
    public RestaurantViewModelFactory(ViewModelSubComponent viewModelSubComponent) {
        creators = new ArrayMap<>();

        // View models cannot be injected directly because they won't be bound to the owner's view model scope.
        creators.put(RestaurantViewModel.class, () -> viewModelSubComponent.restaurantViewModel());
        creators.put(RestaurantListViewModel.class, () -> viewModelSubComponent.restaurantListViewModel());
    }

    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        Callable<? extends ViewModel> creator = creators.get(modelClass);
        if (creator == null) {
            for (Map.Entry<Class, Callable<? extends ViewModel>> entry : creators.entrySet()) {
                if (modelClass.isAssignableFrom(entry.getKey())) {
                    creator = entry.getValue();
                    break;
                }
            }
        }
        if (creator == null) {
            throw new IllegalArgumentException("Unknown model class " + modelClass);
        }
        try {
            return (T) creator.call();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: app/src/main/java/com/navtej/doordashlite/viewmodel/RestaurantViewModelFactory.kt
package com.navtej.doordashlite.viewmodel

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import com.navtej.doordashlite.di.ViewModelSubComponent
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class RestaurantViewModelFactory @Inject constructor(
    private val viewModelSubComponentBuilder: ViewModelSubComponent.Builder
) : ViewModelProvider.Factory {

    private val viewModelSubComponent by lazy {
        viewModelSubComponentBuilder.build()
    }

    private val creators: Map<Class<*>, () -> ViewModel> = mapOf(
        RestaurantViewModel::class.java to { viewModelSubComponent.restaurantViewModel() },
        RestaurantListViewModel::class.java to { viewModelSubComponent.restaurantListViewModel() }
    )

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        val creator = creators[modelClass]
            ?: creators.entries.firstOrNull { modelClass.isAssignableFrom(it.key) }?.value
            ?: throw IllegalArgumentException("Unknown model class $modelClass")
        return creator() as T
    }
}
